src/v/raft/replicate_batcher.cc

```diff
@@ -28,29 +28,33 @@ using namespace std::chrono_literals; // NOLINT
 replicate_batcher::replicate_batcher(consensus* ptr, size_t cache_size)
   : _ptr(ptr)
   , _max_batch_size(cache_size) {
-    _flush_timer.set_callback([this] {
-        (void)ss::with_gate(_ptr->_bg, [this] {
-            // background block further caching too
-            return _lock.with([this] { return flush(); });
-        }).handle_exception_type([this](const ss::gate_closed_exception&) {
-            vlog(
-              _ptr->_ctxlog.debug,
-              "Gate closed while flushing replicate requests");
-        });
+    _flush_timer.set_callback([this] { dispatch_background_flush(); });
+}
+
+void replicate_batcher::dispatch_background_flush() {
+    (void)ss::with_gate(_ptr->_bg, [this] {
+        // background block further caching too
+        return _lock.with([this] { return flush(); });
+    }).handle_exception_type([this](const ss::gate_closed_exception&) {
+        vlog(
+          _ptr->_ctxlog.debug, "Gate closed while flushing replicate requests");
     });
 }
 
 ss::future<result<replicate_result>>
 replicate_batcher::replicate(model::record_batch_reader&& r) {
-    return _lock
-      .with(
-        [this, r = std::move(r)]() mutable { return do_cache(std::move(r)); })
-      .then([this](item_ptr i) {
-          if (_pending_bytes < _max_batch_size || !_flush_timer.armed()) {
-              _flush_timer.rearm(clock_type::now() + 4ms);
-          }
-          return i->_promise.get_future();
-      });
+    return do_cache(std::move(r)).then([this](item_ptr i) {
+        if (_pending_bytes >= _max_batch_size) {
+            _flush_timer.cancel();
+            dispatch_background_flush();
```

Comment:
should `dispatch_background_flush` have an immediate affect which would result in this condition not being true the next time around? what i mean is that i'm guessing `flush` updates `_pending_bytes` but that happens in the background. would we have an issue with many background flushes being dispatched if this code is executed multiple times before the background flush actually runs?

Reply:
since `do_cache` will be executed under the `_lock` the situation described in the example will never take place. Adding new request and flushing will be mutually exclusive.